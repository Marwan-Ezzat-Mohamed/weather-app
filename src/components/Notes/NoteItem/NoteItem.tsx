import React from "react";
import { BiSave, BiEditAlt, BiTrash } from "react-icons/bi";
import { VscDiscard } from "react-icons/vsc";
import { Button } from "@components/common/Button";
import { Textarea } from "@components/common/TextArea";
import { Note } from "@/types";

interface NoteItemProps {
  note: Note;
  onEditNote: (noteId: string, content: string) => void;
  onDeleteNote: (noteId: string) => void;
  onCancelEdit: () => void;
  onUpdateNote: () => void;
  isEditing: boolean;
  editedNote?: string;
  setEditedNote: React.Dispatch<React.SetStateAction<string>>;
}

const NoteItem: React.FC<NoteItemProps> = ({
  note,
  onEditNote,
  onDeleteNote,
  onCancelEdit,
  onUpdateNote,
  setEditedNote,
  isEditing,
  editedNote = "",
}) => {
  return (
    <div
      className="flex w-[calc(50%-4px)] items-end rounded bg-slate-100 p-1 dark:bg-secondary"
      key={note.id}
    >
      <Textarea
        className="w-full resize-none bg-transparent focus-visible:ring-0 focus-visible:ring-offset-0 disabled:text-primary disabled:opacity-60"
        rows={2}
        cols={5}
        id={note.id}
        value={isEditing ? editedNote : note.content}
        disabled={!isEditing}
        onChange={(e) => setEditedNote(e.target.value)}
        data-testid={`textarea-${note.id}`}
      />

      <div>
        {isEditing ? (
          <div className="flex gap-2">
            <Button
              variant="ghost"
              className="p-0 text-3xl"
              onClick={onUpdateNote}
              data-testid={`save-note-${note.id}`}
            >
              <BiSave />
            </Button>
            <Button
              variant="ghost"
              className="p-0 text-3xl"
              onClick={onCancelEdit}
              data-testid={`cancel-edit-${note.id}`}
            >
              <VscDiscard />
            </Button>
          </div>
        ) : (
          <div className="flex gap-2">
            <Button
              variant="ghost"
              className="p-0 text-3xl text-blue-500 dark:text-blue-500 dark:hover:text-blue-700"
              onClick={() => {
                onEditNote(note.id, note.content);
              }}
              data-testid={`edit-note-${note.id}`}
            >
              <BiEditAlt />
            </Button>
            <Button
              variant="ghost"
              className="p-0 text-3xl text-red-500 dark:text-red-500 dark:hover:text-red-700"
              onClick={() => onDeleteNote(note.id)}
              data-testid={`delete-note-${note.id}`}
            >
              <BiTrash />
            </Button>
          </div>
        )}
      </div>
    </div>
  );
};

export default NoteItem;
