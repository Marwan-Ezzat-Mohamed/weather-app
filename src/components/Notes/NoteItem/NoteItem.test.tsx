import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NoteItem from "./NoteItem";

describe("NoteItem", () => {
  const note = {
    id: "1",
    content: "Test note content",
  };

  const onEditNoteMock = jest.fn();
  const onDeleteNoteMock = jest.fn();
  const onCancelEditMock = jest.fn();
  const onUpdateNoteMock = jest.fn();

  it("matches snapshot", () => {
    const { container } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={false}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders note content when not in edit mode", () => {
    const { getByText } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={false}
      />
    );

    const noteContentElement = getByText(note.content);
    expect(noteContentElement).toBeInTheDocument();
  });

  it("renders edit and delete buttons when not in edit mode", () => {
    const { getByTestId } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={false}
      />
    );

    const editButton = getByTestId(`edit-note-${note.id}`);
    const deleteButton = getByTestId(`delete-note-${note.id}`);

    expect(editButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();
  });

  it("calls onEditNote when edit button is clicked", () => {
    const { getByTestId } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={false}
      />
    );

    const editButton = getByTestId(`edit-note-${note.id}`);
    fireEvent.click(editButton);

    expect(onEditNoteMock).toHaveBeenCalledWith(note.id, note.content);
  });

  it("calls onDeleteNote when delete button is clicked", () => {
    const { getByTestId } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={false}
      />
    );

    const deleteButton = getByTestId(`delete-note-${note.id}`);
    fireEvent.click(deleteButton);

    expect(onDeleteNoteMock).toHaveBeenCalledWith(note.id);
  });

  it("renders edit and cancel buttons when in edit mode", () => {
    const { getByTestId } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={true}
      />
    );

    const saveButton = getByTestId(`save-note-${note.id}`);
    const cancelButton = getByTestId(`cancel-edit-${note.id}`);

    expect(saveButton).toBeInTheDocument();
    expect(cancelButton).toBeInTheDocument();
  });

  it("calls onUpdateNote when save button is clicked in edit mode", () => {
    const { getByTestId } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={true}
      />
    );

    const saveButton = getByTestId(`save-note-${note.id}`);
    fireEvent.click(saveButton);

    expect(onUpdateNoteMock).toHaveBeenCalled();
  });

  it("calls onCancelEdit when cancel button is clicked in edit mode", () => {
    const { getByTestId } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={true}
      />
    );

    const cancelButton = getByTestId(`cancel-edit-${note.id}`);
    fireEvent.click(cancelButton);

    expect(onCancelEditMock).toHaveBeenCalled();
  });

  it("displays edited note content when in edit mode", () => {
    const editedNote = "Edited note content";
    const { getByText } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={true}
        editedNote={editedNote}
      />
    );

    const editedNoteContentElement = getByText(editedNote);
    expect(editedNoteContentElement).toBeInTheDocument();
  });

  it("displays the original note content when not in edit mode", () => {
    const editedNote = "Edited note content";
    const { getByText } = render(
      <NoteItem
        note={note}
        onEditNote={onEditNoteMock}
        onDeleteNote={onDeleteNoteMock}
        onCancelEdit={onCancelEditMock}
        onUpdateNote={onUpdateNoteMock}
        setEditedNote={() => {}}
        isEditing={false}
        editedNote={editedNote}
      />
    );

    const noteContentElement = getByText(note.content);
    expect(noteContentElement).toBeInTheDocument();
  });
});
