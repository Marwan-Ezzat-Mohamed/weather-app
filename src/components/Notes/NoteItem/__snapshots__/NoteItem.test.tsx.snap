// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteItem matches snapshot 1`] = `
<div>
  <div
    class="flex w-[calc(50%-4px)] items-end rounded bg-slate-100 p-1 dark:bg-secondary"
  >
    <textarea
      class="placeholder:text-primary/80 flex min-h-[80px] rounded-md px-3 py-2 text-sm text-primary focus-visible:outline-none disabled:cursor-not-allowed w-full resize-none bg-transparent focus-visible:ring-0 focus-visible:ring-offset-0 disabled:text-primary disabled:opacity-60"
      cols="5"
      data-testid="textarea-1"
      disabled=""
      id="1"
      rows="2"
    >
      Test note content
    </textarea>
    <div>
      <div
        class="flex gap-2"
      >
        <button
          class="inline-flex items-center justify-center rounded-md font-medium transition-colors focus:outline-none focus:ring-slate-400 disabled:opacity-50 dark:focus:ring-slate-400 disabled:pointer-events-none dark:focus:ring-offset-primary focus:ring-0 focus:ring-offset-0 ring-0 bg-transparent hover:bg-slate-100 dark:hover:bg-secondary data-[state=open]:bg-transparent dark:data-[state=open]:bg-transparent h-10 p-0 text-3xl text-blue-500 dark:text-blue-500 dark:hover:text-blue-700"
          data-testid="edit-note-1"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.045 7.401c.378-.378.586-.88.586-1.414s-.208-1.036-.586-1.414l-1.586-1.586c-.378-.378-.88-.586-1.414-.586s-1.036.208-1.413.585L4 13.585V18h4.413L19.045 7.401zm-3-3 1.587 1.585-1.59 1.584-1.586-1.585 1.589-1.584zM6 16v-1.585l7.04-7.018 1.586 1.586L7.587 16H6zm-2 4h16v2H4z"
            />
          </svg>
        </button>
        <button
          class="inline-flex items-center justify-center rounded-md font-medium transition-colors focus:outline-none focus:ring-slate-400 disabled:opacity-50 dark:focus:ring-slate-400 disabled:pointer-events-none dark:focus:ring-offset-primary focus:ring-0 focus:ring-offset-0 ring-0 bg-transparent hover:bg-slate-100 dark:hover:bg-secondary data-[state=open]:bg-transparent dark:data-[state=open]:bg-transparent h-10 p-0 text-3xl text-red-500 dark:text-red-500 dark:hover:text-red-700"
          data-testid="delete-note-1"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 20a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V8h2V6h-4V4a2 2 0 0 0-2-2H9a2 2 0 0 0-2 2v2H3v2h2zM9 4h6v2H9zM8 8h9v12H7V8z"
            />
            <path
              d="M9 10h2v8H9zm4 0h2v8h-2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
